lib/ddtrace/sampling/rule_sampler.rb

```diff
@@ -0,0 +1,77 @@
+require 'forwardable'
+
+require 'ddtrace/ext/priority'
+
+require 'ddtrace/ext/sampling'
+require 'ddtrace/sampler'
+require 'ddtrace/sampling/token_bucket'
+
+module Datadog
+  module Sampling
+    # TODO: Write class documentation
+    # RuleSampler
+    class RuleSampler
+      extend Forwardable
+
+      attr_reader :rules, :rate_limiter, :fallback_sampler
+
+      def initialize(rules = [],
+                     rate_limiter = Datadog::Sampling::UnlimitedLimiter.new,
+                     fallback_sampler = Datadog::AllSampler.new)
+        @rules = rules
+        @rate_limiter = rate_limiter
+        @fallback_sampler = fallback_sampler
+      end
+
+      # /RuleSampler's components (it's rate limiter, for example) are
+      # not be guaranteed to be size-effect free.
+      # It is not possible to guarantee that a call to {#sample?} will
+      # return the same result as a successive call to {#sample!} with the same span.
+      #
+      # Use {#sample!} instead
+      def sample?(_span)
+        raise 'RuleSampler cannot be evaluated without side-effects'
+      end
+
+      def sample!(span)
+        sampled = sample_span(span) { |s| @fallback_sampler.sample!(s) }
+
+        sampled.tap do
+          span.sampled = sampled
```

Comment:
for here, we always want `span.sampled = true`, we want to do:

```ruby
span.sampled = true
span.context.sampling_priority = sampled ? Datadog::Ext::Priority::AUTO_KEEP : Datadog::Ext::Priority::AUTO_REJECT
```

Reply:
Our samplers can be used as stand-alone samplers, as well as composed together.
When using `RuleSampler` as a stand-alone sampler, it will make hard decisions.

When using it alongside the `PrioritySampler`, the sampling_priority and final sampling decision are taken care of by the priority sampler, which will enforce the conditions you mentioned.

`PrioritySampler` is currently the sampler in our code base that is aware of `sampling_priority`. All other samplers only make a boolean decision and are not concerned with details around what flags to set in the `span`.